### ISSUE TITLE
Prevent "bad" shard from causing perpetual initialization tasks

### ISSUE THREAD

**Issue by ppf2** [elastic/elasticsearch#11309]
Have come across this multiple times in the field.  For whatever reason (still to be determined - different issue), the index files for a shard are gone (for both the primary and replica).  The shard's index directory exists (but is empty) for the shard for both primary and replica.  When this occurs:

* If the end user checks the cluster health api, it will show that there is a shard initializing, but it will be like this forever as if it is taking a long time to initialize.  But in fact, it is in initializing state because it keeps trying and never stops trying.
* If you check the cat/shards, you will see that the shard keeps trying to initializing on the 2 nodes, basically, initialize on node 1 -> fails -> initialize on node 2 -> fails, initialize on node 1 -> fails, etc.. and so on.  If you look at head, you will see that the shard is "bouncing" between the 2 nodes:

![image](https://cloud.githubusercontent.com/assets/7216393/7776570/a8af3e76-0070-11e5-94c4-79b2b95b8857.png)

![image](https://cloud.githubusercontent.com/assets/7216393/7776564/a136a1a2-0070-11e5-839f-266f5aa9dfd8.png)

* And corresponding pending tasks keep showing up for each attempt against the 2 nodes:

```
{
  "tasks" : [ {
    "insert_order" : 14347,
    "priority" : "HIGH",
    "source" : "shard-failed ([default][0], node[5xR9MWlmSDKHjZPIZs4L-w], [P], s[INITIALIZING]), reason [Failed to start shard, message [IndexShardGatewayRecoveryException[[default][0] failed to fetch index version after copying it over]; nested: IndexShardGatewayRecoveryException[[default][0] shard allocated for local recovery (post api), should exist, but doesn't, current files: [.DS_Store]]; nested: IndexNotFoundException[no segments* file found in store(least_used[rate_limited(niofs(/Users/ppf2/ELK/elasticsearch-1.4.2_1/data/elasticsearch_1_4_2/nodes/0/indices/default/0/index), type=MERGE, rate=20.0)]): files: [.DS_Store]]; ]]",
    "executing" : true,
    "time_in_queue_millis" : 13,
    "time_in_queue" : "13ms"
  } ]
}
```

* And the logs on both nodes will start filling up with a ton of these warning messages:

```
[2015-05-22 10:56:27,812][WARN ][cluster.action.shard     ] [node2] [default][0] sending failed shard for [default][0], node[ktzJrukQQDux_edYck2tYA], [P], s[INITIALIZING], indexUUID [zdMk9SthR_OzhCJi27QzcQ], reason [Failed to start shard, message [IndexShardGatewayRecoveryException[[default][0] failed to fetch index version after copying it over]; nested: IndexShardGatewayRecoveryException[[default][0] shard allocated for local recovery (post api), should exist, but doesn't, current files: [.DS_Store]]; nested: IndexNotFoundException[no segments* file found in store(least_used[rate_limited(niofs(/Users/ppf2/ELK/elasticsearch-1.4.2_2/data/elasticsearch_1_4_2/nodes/0/indices/default/0/index), type=MERGE, rate=20.0)]): files: [.DS_Store]]; ]]
[2015-05-22 10:56:27,831][WARN ][indices.cluster          ] [node2] [default][0] failed to start shard
org.elasticsearch.index.gateway.IndexShardGatewayRecoveryException: [default][0] failed to fetch index version after copying it over
	at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:158)
	at org.elasticsearch.index.gateway.IndexShardGatewayService$1.run(IndexShardGatewayService.java:132)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.elasticsearch.index.gateway.IndexShardGatewayRecoveryException: [default][0] shard allocated for local recovery (post api), should exist, but doesn't, current files: [.DS_Store]
	at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:132)
	... 4 more
Caused by: org.apache.lucene.index.IndexNotFoundException: no segments* file found in store(least_used[rate_limited(niofs(/Users/ppf2/ELK/elasticsearch-1.4.2_2/data/elasticsearch_1_4_2/nodes/0/indices/default/0/index), type=MERGE, rate=20.0)]): files: [.DS_Store]
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:864)
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:752)
	at org.apache.lucene.index.SegmentInfos.read(SegmentInfos.java:450)
	at org.elasticsearch.common.lucene.Lucene.readSegmentInfos(Lucene.java:85)
	at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:123)
	... 4 more
```

To reproduce:
* Have a cluster with 2 node and an index with 1 replica.   Go to the file system and simulate the loss of shard data by deleting all files under a shard's index folder (keep the directory but empty it) from both nodes.
* Restart the 2 nodes
* Then you can see the continuous attempts to start the shard on both nodes and it will keep trying and keep failing and doesn't give up.

I suspect that the same behavior will occur if both primary and replica shards have index corruptions.

It will be nice to:

* Prevent it from trying perpetually - once it determines that there are no copies of the shard useable/available, stop trying repeatedly on the nodes
* Leave the problem shard unassigned and write an intuitive message to the master log with a summary of what nodes it tried on and that it is leaving the shard unassigned because it cannot find a usable copy of the shard data, etc..

**Comment by s1monw** [comment]
this is fixed by #11269 as of yesterday :) 

it will now try once this:

```
[2015-05-22 20:48:46,071][DEBUG][index                    ] [Erik Josten] [foo] [0] closed (reason: [failed recovery])
[2015-05-22 20:48:46,072][WARN ][indices.cluster          ] [Erik Josten] [[foo][0]] marking and sending shard failed due to [failed recovery]
org.elasticsearch.index.gateway.IndexShardGatewayRecoveryException: [foo][0] failed to fetch index version after copying it over
	at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:158)
	at org.elasticsearch.index.gateway.IndexShardGatewayService$1.run(IndexShardGatewayService.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.elasticsearch.index.gateway.IndexShardGatewayRecoveryException: [foo][0] shard allocated for local recovery (post api), should exist, but doesn't, current files: []
	at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:132)
	... 4 more
Caused by: org.apache.lucene.index.IndexNotFoundException: no segments* file found in store(least_used[rate_limited(niofs(/Users/simon/projects/elasticsearch/target/releases/elasticsearch-1.6.0-SNAPSHOT/data/elasticsearch/nodes/0/indices/foo/0/index), type=MERGE, rate=20.0)]): files: []
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:881)
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:769)
	at org.apache.lucene.index.SegmentInfos.read(SegmentInfos.java:458)
	at org.elasticsearch.common.lucene.Lucene.readSegmentInfos(Lucene.java:89)
	at org.elasticsearch.index.store.Store.readSegmentsInfo(Store.java:166)
	at org.elasticsearch.index.store.Store.readLastCommittedSegmentsInfo(Store.java:152)
	at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:123)
	... 4 more
```

then log this. I think we should upgrade this to a warn logging @kimchy WDYT?
```
[2015-05-22 20:48:46,135][DEBUG][gateway.local            ] [Erik Josten] [foo][0] found 0 allocations of [foo][0], node[null], [P], s[UNASSIGNED], highest version: [-1]

```

**Comment by kimchy** [comment]
@s1monw the tricky bit on the logging is that this might be valid, as in, when you start nodes one by one on a fresh cluster, you will get to this state, which might be valid. I wonder if the indication of shard failures is enough to indicate there is a problem with this shard if its left unassigned? There is also an option of making it info.

**Comment by s1monw** [comment]
ok fair enough

**Comment by ppf2** [comment]
Thanks guys!

**Comment by s1monw** [comment]
just for the record this is fixed in `1.6` & `2.0`
